test/helpers/position_assertions.cc

```diff
@@ -1152,24 +1154,129 @@ shared_ptr<SymbolSearchAssertion> SymbolSearchAssertion::make(string_view filena
+    }
+    if (name.has_value() && name.value() != symbol->name) {
+        return false;
+    }
+    return true;
+}
+namespace {
+string pluralized_count(string_view word, int count) {
+    return fmt::format("{} {}{}", count, word, (count == 1) ? "" : "s");
+}
 
-    std::map<std::string, vector<shared_ptr<SymbolSearchAssertion>>> assertionsByLine;
+void addFailureAtLocationWithSource(unique_ptr<Location> location,
+                                    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents,
+                                    string_view uriPrefix, std::string_view message) {
+    auto sourceLine = getLine(sourceFileContents, uriPrefix, *location);
+    ADD_FAILURE_AT(location->uri.c_str(), location->range->start->line + 1)
+        << fmt::format("{}\n{}\n", message, prettyPrintRangeComment(sourceLine, *location->range, ""));
+}
+
+/*
+ * Splits assertions/symbols for a query into three categories:
+ *    assertionSymbolPairs => matching pairs of assertions/symbols
+ *    unmatchedAssertions => assertions with no matching symbol
+ *    unmatchedSymbols => symbols with no matching assertion
+ */
+void matchAssertionsToSymbols(
+    string_view query, const vector<shared_ptr<SymbolSearchAssertion>> &assertions,
+    vector<shared_ptr<SymbolInformation>> &symbols,
+    const UnorderedMap<string, shared_ptr<core::File>> &sourceFileContents, string_view uriPrefix,
+    string_view errorPrefix,
+    vector<pair<shared_ptr<SymbolSearchAssertion>, shared_ptr<SymbolInformation>>> &assertionSymbolPairs,
+    vector<shared_ptr<SymbolSearchAssertion>> &unmatchedAssertions,
+    vector<shared_ptr<SymbolInformation>> &unmatchedSymbols) {
+    std::map<std::pair<std::string, int>, vector<shared_ptr<SymbolSearchAssertion>>> assertionsByLine;
     for (auto assertion : assertions) {
         auto location = assertion->getLocation(uriPrefix);
-        auto line = fmt::format("{}:{:{}}", location->uri, location->range->start->line, lineWidth);
-        assertionsByLine[line].push_back(assertion);
+        assertionsByLine[make_pair(location->uri, location->range->start->line)].push_back(assertion);
+    }
+    std::map<std::pair<std::string, int>, vector<shared_ptr<SymbolInformation>>> symbolsByLine;
+    for (auto &symbol : symbols) {
+        auto key = make_pair(symbol->location->uri, symbol->location->range->start->line);
+        symbolsByLine[key].push_back(symbol);
+        assertionsByLine[key]; // create entry in assertionsByLine, if none exists yet.
+    }
+    // Look line-by-line for unmatched symbols and assertions
+    for (auto entry : assertionsByLine) {
+        auto assertionsOnLine = entry.second;
+        auto &symbolsOnLine = symbolsByLine[entry.first];
+        auto [uri, line] = entry;
+        // Find assertions with no matching symbol
+        for (auto &assertion : assertionsOnLine) {
+            optional<shared_ptr<SymbolInformation>> matchingSymbol;
+            for (auto &symbol : symbols) {
```

Comment:
Since some symbols are coming from `rbi` files, it's now the case that assertions can match symbols that aren't on the same line, so now we *are* doing <# symbols x # assertions> checks, but I hope/think test behavior is easier to understand because of it.